### DISCUSSION TITLE
Throughput Testing

### DISCUSSION THREAD

**Discussion by Oles7** [jmmorato/openddsharp#230] [General]
Are there any tests to measure the throughput of OpenDDSharp?

I am trying to measure throughput by publishing a topic like:

@topic
struct{
   char arr[8192]
}

I am using a for loop to continuously publish this topic for 10 seconds on the publisher side.

The subscriber side counts all the messages it receives, the start time and the end time and then calculates the throughput by diving the total number of bits it received  (8192 bytes * 8 and then multiplied by the number of messages it received) by the number of seconds it took.

Are there any qos options that can be used to optimize the throughput?
It looks like the publish method for the topic above takes around 5.6ms for every publish which winds up getting a throughput of about 14.5 Mbits/second.

We need to achieve a throughput of 100 Mbits/second.

**Comment by jmmorato** [comment]
> Are there any tests to measure the throughput of OpenDDSharp?
 
 No, there is no performance tests available for networking yet but an initial benchmark performance project was pushed some month ago.
 
 > Are there any qos options that can be used to optimize the throughput?
 
 Surely a QoS refinement could help you to improve a bit but the main bottleneck right now is the serialization/deseralization of the samples to interop with the C++ OpenDDS layer. A new CDR serialization/deseralization version is coming to workaround that issue but still in development.
 
 > We need to achieve a throughput of 100 Mbits/second.
 
 If you share your code, I could try to help you to improve your performance

**Reply by Oles7** [reply]
I cloned the openddsharp git repository to virtual studio and it looks like it tries to build but it cannot find dependencies for OpenDds. It finds the OpenDdsConfig.cmake file but cannot find OPENDDS_DCPS_LIBRARY and _OPENDDS_OPENDDS_OPENDDS_IDL. Any ideas how I can fix that?


t seems like the EncodeToString function in Write is what is causing it to slow down. Is that the serialization part? Are there any types we could use in the .idl file to skip or optimize that part? We've gotten the best results using a Boolean array of size 64000. That got our throughput up to 40Mbits/second. We're still stuck on how to improve it. I can share the code later. We are using best effort qos. In the rtps file we use tcp.

**Comment by jmmorato** [comment]
> I cloned the openddsharp git repository to virtual studio and it looks like it tries to build but it cannot find dependencies for OpenDds

No documentation on how to build yet... 🤦‍♂️ Best is to check how OpenDDSharp is build in the CI/CD pipelines

https://github.com/jmmorato/openddsharp/blob/0d1c5732772e66417f0d886ea385f97ecce948b3/.github/workflows/ci_standard.yaml#L86C5-L91C1

> Is that the serialization part?

Exactly, that's my point. The JSON serialisation is not good enough for throughput performance that's why I'm working on the CDR serialisation in parallel. Another solution could be using Dynamic Data but I don't even started the development. In future version, the three solutions will coexist and the developer will use the better it fits for his solution.

> Are there any types we could use in the .idl file to skip or optimize that part?

I would put my two cents in CDR serialisation with a byte array (it will require a beta version with the current develop branch); bytes arrays are easy for memory management.
